test/Microsoft.AspNet.Razor.Test/TagHelpers/TagHelperParseTreeRewriterTest.cs

```diff
@@ -19,6 +19,877 @@ namespace Microsoft.AspNet.Razor.Test.TagHelpers
+                                    factory.Markup("btn").With(
+                                        new LiteralAttributeCodeGenerator(
+                                            prefix: new LocationTagged<string>(string.Empty, 11, 0, 11),
+                                            value: new LocationTagged<string>("btn", 11, 0, 11))),
+                                    factory.Markup("\"").With(SpanCodeGenerator.Null))))
+                    }
+                };
+            }
+        }
+
+        [Theory]
+        [MemberData(nameof(OptOut_WithPartialData_HTML))]
+        public void Rewrite_AllowsTagHelperElementOptForIncompleteHTML(
+            string documentContent,
+            MarkupBlock expectedOutput)
+        {
+            RunParseTreeRewriterTest(documentContent, expectedOutput, new RazorError[0], "strong", "p");
+        }
+
+        public static TheoryData OptOut_WithBlockData_CSharp
+        {
+            get
+            {
+                var factory = CreateDefaultSpanFactory();
+                var blockFactory = new BlockFactory(factory);
+                var errorFormatMalformed =
+                    "Found a malformed '{0}' tag helper. Tag helpers must have a start and end tag or be self " +
+                    "closing.";
+                var errorFormatNormalUnclosed =
+                    "The \"{0}\" element was not closed.  All elements must be either self-closing or have a " +
+                    "matching end tag.";
+                var errorFormatNormalNotStarted =
+                    "Encountered end tag \"{0}\" with no matching start tag.  Are your start/end tags properly " +
+                    "balanced?";
+                var errorMatchingBrace =
+                    "The code block is missing a closing \"}\" character.  Make sure you have a matching \"}\" " +
+                    "character for all the \"{\" characters within this block, and that none of the \"}\" " +
+                    "characters are being interpreted as markup.";
+
+                Func<Func<MarkupBlock>, MarkupBlock> buildStatementBlock = (insideBuilder) =>
+                {
+                    return new MarkupBlock(
+                        factory.EmptyHtml(),
+                        new StatementBlock(
+                            factory.CodeTransition(),
+                            factory.MetaCode("{").Accepts(AcceptedCharacters.None),
+                            insideBuilder(),
+                            factory.EmptyCSharp().AsStatement(),
+                            factory.MetaCode("}").Accepts(AcceptedCharacters.None)),
+                        factory.EmptyHtml());
+                };
+
+                // documentContent, expectedOutput, expectedErrors
+                return new TheoryData<string, MarkupBlock, RazorError[]>
+                {
+                    {
+                        "@{<!p></!p>}",
```

Comment:
- add cases like `@{<text></!text>}` too
- should also cover one or two cases w/ a regular span between mismatched tags e.g. `@{<!p>words and spaces</!p>}`